core/src/main/scala-2/chisel3/Contract.scala

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import chisel3.experimental.SourceInfo
+import chisel3.internal._
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl.ir._
+import scala.language.experimental.macros
+import scala.reflect.macros.whitebox
+
+object Contract {
+
+  /** Create a `contract` block with no arguments and results.
+    *
+    * Within the contract body, `RequireProperty` and `EnsureProperty` can be
+    * used to describe the pre and post-conditions of the contract. During
+    * formal verification, contracts can be used to divide large formal checks
+    * into smaller pieces by first asserting some local properties hold, and
+    * then assuming those properties hold in the context of the larger proof.
+    *
+    * @example
+    * {{{
+    * Contract() {
+    *   RequireProperty(a >= 1)
+    *   EnsureProperty(a + a >= 2)
+    * }
+    * }}}
+    */
+  def apply()(body: => Unit)(implicit sourceInfo: SourceInfo): Unit = {
+    mapped[Unit](Seq(), _ => ())(_ => body)
+  }
+
+  /** Create a `contract` block with one or more arguments and results.
+    *
+    * Outside of verification uses, the arguments passed to the contract are
+    * simply returned as contract results. During formal verification, the
+    * contract may act as a cutpoint by returning symbolic values for its
+    * results and using the contract body as constraint.
+    *
+    * Within the contract body, `RequireProperty` and `EnsureProperty` can be
+    * used to describe the pre and post-conditions of the contract. During
+    * formal verification, contracts can be used to divide large formal checks
+    * into smaller pieces by first asserting some local properties hold, and
+    * then assuming those properties hold in the context of the larger proof.
+    *
+    * @example
+    * {{{
+    * // Contract with single argument: "(a<<3)+a is equivalent to a*9".
+    * // During formal verification, this is interpreted as:
+    * // - "assert((a<<3)+a === a*9)" as a proof that the contract holds
+    * // - "assume(b === a*9)" as a simplification of other proofs
+    * val b = Contract((a<<3)+a) { b =>
+    *   EnsureProperty(b === a*9)
+    * }
+    *
+    * // Contract with multiple arguments:
+    * //   "carry-save adder reduces number of terms but doesn't change sum"
+    * // During formal verification, this is interpreted as:
+    * // - an assertion that the carry-save adder indeed compresses the three
+    * //   terms p/q/r down to only two terms u/v which have the same sum
+    * // - symbolic values u/v constrained to sum up to the same value as p/q/r,
+    * //   as a simplification of other proofs
+    * val csa_s = p^q^r
+    * val csa_c = (p&q | (p^q)&r) << 1
+    * val (u,v) = Contract(csa_c, csa_s) { case (u,v) =>
```

Comment:
again, what is the point of `val (u, v)` that is returned here? what are they and what do you do with them?